### PR TITLE
Stabilize support for Retry-After header (MSC4014)

### LINKED ISSUE #16944: Stabilize MSC4041 support (Retry-After header)
[MSC4041](https://github.com/matrix-org/matrix-spec-proposals/pull/4041) has passed FCP and can be stabilized in Synapse.

See matrix-org/synapse#16136 for the initial implementation.

The steps to do this are:

1. Remove the changes in `synapse/config/experimental.py` to set `LimitExceededError.include_retry_after_header`.
2. Remove the `LimitExceededError.include_retry_after_header` flag (and default to it being enabled always).
3. Update the tests at `tests/api/test_errors.py` and `tests/rest/client/test_login.py`.

That should be all that's necessary, but some other tests could break due to this. It should be a fully backwards compatible change.

### PR DESCRIPTION
**Pull request by clokep**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Fixes #16944

### CODE REVIEW

File: synapse/api/errors.py [hunk 1]
```diff
@@ -526,9 +526,10 @@ def __init__(
         retry_after_ms: Optional[int] = None,
         errcode: str = Codes.LIMIT_EXCEEDED,
     ):
+        # Use HTTP header Retry-After to enable library-assisted retry handling.
         headers = (
             {"Retry-After": str(math.ceil(retry_after_ms / 1000))}
-            if self.include_retry_after_header and retry_after_ms is not None
```
Comment (anoadragon453) [inline synapse/api/errors.py:531]: `self.include_retry_after_header` can now be deleted.

File: tests/api/test_errors.py [hunk 2]
```diff
@@ -33,18 +33,14 @@ def test_key_appears_in_context_but_not_error_dict(self) -> None:
         self.assertIn("needle", err.debug_context)
         self.assertNotIn("needle", serialised)
 
-    # Create a sub-class to avoid mutating the class-level property.
-    class LimitExceededErrorHeaders(LimitExceededError):
-        include_retry_after_header = True
-
     def test_limit_exceeded_header(self) -> None:
-        err = self.LimitExceededErrorHeaders(limiter_name="test", retry_after_ms=100)
+        err = LimitExceededError(limiter_name="test", retry_after_ms=100)
```
Comment (anoadragon453) [inline tests/api/test_errors.py:37]: Seems there are a few instances of `msc4041_enabled` still being set in the tests, yet the option no longer exists.

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
A couple more cleanups to do, but otherwise LGTM.

Thanks for doing this, and for adding a clarifying comment!

**Comment by clokep** [comment]
This should be ready for another review, assuming CI passes.

**Review by anoadragon453 — APPROVED** [review]
LGTM. Thank you very much!

CI failures indicate a flaky sytest.

**Pull request merged**